src/Extension/Propagator/XCloudTrace/XCloudTraceFormatter.php

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace OpenTelemetry\Extension\Propagator\XCloudTrace;
+
+use OpenTelemetry\API\Trace\SpanContext;
+use OpenTelemetry\API\Trace\SpanContextInterface;
+
+/**
+ * This format using a human readable string encoding to propagate SpanContext.
+ * The current format of the header is `<trace-id>[/<span-id>][;o=<options>]`.
+ * The options are a bitmask of options. Currently the only option is the
+ * least significant bit which signals whether the request was traced or not
+ * (1 = traced, 0 = not traced).
+ */
+final class XCloudTraceFormatter
+{
+    const CONTEXT_HEADER_FORMAT = '/([0-9a-fA-F]{32})(?:\/(\d+))?(?:;o=(\d+))?/';
+
+    /**
+     * Generate a SpanContext object from the Trace Context header
+     *
+     * @param string $header
+     * @return SpanContextInterface
+     */
+    public static function deserialize(string $header) : SpanContextInterface
+    {
+        if (preg_match(self::CONTEXT_HEADER_FORMAT, $header, $matches)) {
+            return SpanContext::createFromRemoteParent(
+                strtolower($matches[1]),
+                array_key_exists(2, $matches) && !empty($matches[2])
+                    ? self::decToHex($matches[2])
+                    : null,
+                array_key_exists(3, $matches)
+                    ? (int)($matches[3] == '1')
+                    : null
+            );
+        }
+        return SpanContext::getInvalid();
+    }
+
+    /**
+     * Convert a SpanContextInterface to header string
+     *
+     * @param SpanContextInterface $context
+     * @return string
+     */
+    public static function serialize(SpanContextInterface $context) : string
+    {
+        $ret = $context->getTraceId();
+        if ($context->getSpanId()) {
+            $ret .= '/' . self::hexToDec($context->getSpanId());
+        }
+        $ret .= ';o=' . ($context->isSampled() ? '1' : '0');
+        return $ret;
+    }
+
+    private static function decToHex(string $num) : string
+    {
+        $int = (int) $num;
+        if (self::isBigNum($int)) {
+            $ret = self::baseConvert($num, 10, 16);
+        } else {
+            $ret = dechex($int);
+        }
+        return str_pad($ret, 16, '0', STR_PAD_LEFT);
+    }
+
+    private static function hexToDec(string $num) : string
+    {
+        $dec = hexdec($num);
+        if (self::isBigNum($dec)) {
+            return self::baseConvert($num, 16, 10);
+        }
+        return strval($dec);
+    }
+
+    private static function isBigNum(int|float $number) : bool
+    {
+        return $number >= PHP_INT_MAX;
+    }
+
+    private static function baseConvert(string $num, int $fromBase, int $toBase) : string
```

Comment:
this seems generally useful, and under-documented. could it for now be moved into its own class and unit tested?

Reply:
Sure. For reference, this function covers 2 cases:this code makes octal->int->str work on 32 bit systems (which will rarely ever happen), and when the span id is over the signed max long limit, but under the unsigned max long limit. PHP does not easily support unsigned longs, so this function is necessary. This will definitely take a bit of reference engineering.

Reply:
Updated, repushed PR.